src/video/wayland/SDL_waylandevents.c

```diff
@@ -1725,20 +1732,52 @@ static void data_device_handle_drop(void *data, struct wl_data_device *wl_data_d
     if (data_device->drag_offer != NULL) {
         /* TODO: SDL Support more mime types */
         size_t length;
-        void *buffer = Wayland_data_offer_receive(data_device->drag_offer,
-                                                  &length, FILE_MIME, SDL_TRUE);
-        if (buffer) {
-            char *saveptr = NULL;
-            char *token = SDL_strtokr((char *)buffer, "\r\n", &saveptr);
-            while (token != NULL) {
-                char *fn = Wayland_URIToLocal(token);
-                if (fn) {
-                    SDL_SendDropFile(data_device->dnd_window, fn);
+        SDL_bool drop_handled = SDL_FALSE;
+#ifdef SDL_USE_LIBDBUS
+        if (Wayland_data_offer_has_mime(
```

Comment:
why doesn't the older SDL_strcmp approach work here?

Reply:
If you mean [this use of SDL_strcmp](https://github.com/libsdl-org/SDL/commit/127bb9f15ea6759deea93c4692868703f217d04c#diff-8a9cade1c69cd3c5c95378f2b386222c68aa893ea73ae792bb865061b8b9ebf2R1739), I changed `handle_enter` to `wl_data_offer_accept` both supported mime types which made `accepted_mime` no longer meaningful as a string.

Reply:
> If you mean [this use of SDL_strcmp](https://github.com/libsdl-org/SDL/commit/127bb9f15ea6759deea93c4692868703f217d04c#diff-8a9cade1c69cd3c5c95378f2b386222c68aa893ea73ae792bb865061b8b9ebf2R1739), I changed `handle_enter` to `wl_data_offer_accept` both supported mime types which made `accepted_mime` no longer meaningful as a string.

yes but doing it that way seems much more complicated than the strcmp method

Reply:
What strings are you comparing in your case with fallback?

Reply:
> What strings are you comparing in your case with fallback?

the same strings you were comparing, FILE_PORTAL_MIME and FILE_MIME

Reply:
Against a string `accepted_mime` set on `handle_enter`?

Reply:
If SDL `wl_data_offer_accept`s both `text/url-list` and `application/vnd.portal.filetransfer`, we could keep a list of types.
Out of interest I've been testing this code with:
```bash
cmake --build build && flatpak run --filesystem="$(pwd)" --socket=wayland org.freedesktop.Platform//23.08 -c ./build/test/testdropfile
```

Reply:
> Against a string `accepted_mime` set on `handle_enter`?

accepted_mime

Reply:
> If SDL `wl_data_offer_accept`s both `text/url-list` and `application/vnd.portal.filetransfer`, we could keep a list of types. Out of interest I've been testing this code with:
> 
> ```shell
> cmake --build build && flatpak run --filesystem="$(pwd)" --socket=wayland org.freedesktop.Platform//23.08 -c ./build/test/testdropfile
> ```

doesn't each drag offer only have one type, as far as I know you can't have 2 types on a single drag offer?

Reply:
No, the `data_offer_handle_offer` function is called for each mime type for a given `wl_data_offer` https://wayland.app/protocols/wayland#wl_data_offer:event:offer

Reply:
So for a pending file drop on mutter in flatpak I get: 
```
text/plain
text/plain;charset=utf-8
application/x-gtk-local-dnd
application/vnd.portal.filetransfer
application/vnd.portal.files
text/uri-list
text/plain;charset=utf-8
```

Reply:
yes, but each offer only has one type, and we only want to accept one of those so that we don't receive 10 files for each drag and drop

Reply:
The mime type doesn't represent that of the file.
It represents the type of data to be transferred via [`wl_data_offer_receive`](https://wayland.app/protocols/wayland#wl_data_offer:request:receive)

Reply:
So for example `uri-list` will be a list of dropped files using the `file://` uri scheme.
And `text/plain` may be a list of simple paths strings.
I believe `application/vnd.portal.files` is deprecated name for `application/vnd.portal.filetransfer` used for backwards compatibility.

Reply:
So we accept uri-list and vnd.portal so that we can fallback to uri-list when vnd.portal fails.

Reply:
> So we accept `uri-list` and `vnd.portal` so that we can fallback to `uri-list` when `vnd.portal` fails.

ok I understand now thanks